libs/collectives/include/hpx/collectives/scatter.hpp

```diff
@@ -0,0 +1,516 @@
+            hpx::id_type id = fid.get();
+            auto result = async(action_type(), id, this_site);
+
+            traits::detail::get_shared_state(result)->set_on_completed(
+                [id = std::move(id)]() { HPX_UNUSED(id); });
+
+            return result;
+        };
+
+        return fid.then(hpx::launch::sync, std::move(scatter_data));
+    }
+
+    template <typename T>
+    hpx::future<T> scatter_from(char const* basename,
+        std::size_t generation = std::size_t(-1),
+        std::size_t this_site = std::size_t(-1), std::size_t root_site = 0)
+    {
+        if (this_site == std::size_t(-1))
+        {
+            this_site = static_cast<std::size_t>(hpx::get_locality_id());
+        }
+
+        std::string name(basename);
+        if (generation != std::size_t(-1))
+        {
+            name += std::to_string(generation) + "/";
+        }
+
+        return scatter_from<T>(
+            hpx::find_from_basename(std::move(name), root_site), this_site);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // scatter_to
+    template <typename T>
+    hpx::future<T> scatter_to(hpx::future<hpx::id_type>&& fid,
+        hpx::future<std::vector<T>>&& result,
+        std::size_t this_site = std::size_t(-1))
+    {
+        if (this_site == std::size_t(-1))
+        {
+            this_site = static_cast<std::size_t>(hpx::get_locality_id());
+        }
+
+        auto scatter_to_data =
+            [this_site](hpx::future<hpx::id_type>&& fid,
+                hpx::future<std::vector<T>>&& local_result) -> hpx::future<T> {
+            using action_type = typename detail::communicator_server<T>::
+                template communication_set_action<
+                    traits::communication::scatter_tag, T, std::vector<T>>;
+
+            // make sure id is kept alive as long as the returned future
+            hpx::id_type id = fid.get();
+            auto result =
+                async(action_type(), id, this_site, local_result.get());
+
+            traits::detail::get_shared_state(result)->set_on_completed(
+                [id = std::move(id)]() { HPX_UNUSED(id); });
+
+            return result;
+        };
+
+        return dataflow(hpx::launch::sync, std::move(scatter_to_data),
+            std::move(fid), std::move(result));
+    }
+
+    template <typename T>
+    hpx::future<T> scatter_to(char const* basename,
+        hpx::future<std::vector<T>>&& result,
+        std::size_t num_sites = std::size_t(-1),
+        std::size_t generation = std::size_t(-1),
+        std::size_t this_site = std::size_t(-1))
+    {
+        if (num_sites == std::size_t(-1))
+        {
+            num_sites = static_cast<std::size_t>(
+                hpx::get_num_localities(hpx::launch::sync));
+        }
+        if (this_site == std::size_t(-1))
+        {
+            this_site = static_cast<std::size_t>(hpx::get_locality_id());
+        }
```

Comment:
Is there a reason these can't just directly be the default parameters? I suppose to be able to specify `generation` but leave `num_sites` default? Related question, but definitely not for this PR: should `get_num_localities` and the locality id just be `std::size_t`?

Reply:
> Is there a reason these can't just directly be the default parameters? I suppose to be able to specify generation but leave num_sites default?

Yes, this is fixed by #4687 (which relies on this PR).

> Related question, but definitely not for this PR: should get_num_localities and the locality id just be std::size_t?

Ok, not objections. See #4697.